### PR TITLE
fix(vrl): fix if statement type definitions

### PR DESCRIPTION
**Pull request by fuchsnj**
fixes: https://github.com/vectordotdev/vector/issues/11773
fixes: https://github.com/vectordotdev/vector/issues/12948

- This now correctly calculates the type definitions of locals and external values in if/else statements.
- The expression type of an if statement without an else clause now includes `null`. The actual value could already be null, but the type definition didn't include it.

### Examples

Just if statement (original and block type are merged)

```
$ a = 5
5

$ if false {a = "string"}
null

$ a
5

$ type_def(a)
{ "bytes": true, "integer": true }
```

If/else statement (only the if/else types are merged)

```
$ a = 5
5

$ if false {a = "string"} else {a = {}}
{  }

$ type_def(a)
{ "bytes": true, "object": {  } }
```

### CODE REVIEW

File: lib/vrl/tests/tests/expressions/if_statement/if_else_external_assignment.vrl [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# result: true
+
+.a = 5
+if true {
+  .a = 0.0
+} else {
+  .a = "string"
+}
+assert!(.a == 0.0)
+assert!(type_def(.a) == {"float": true, "bytes": true})
```
Comment (jszwedko) [inline lib/vrl/tests/tests/expressions/if_statement/if_else_external_assignment.vrl:10]: 👍 being able to assert the type def is useful.

File: lib/vrl/tests/tests/issues/12948_conditional_type_def_regression.vrl [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# result: null
+
+if .platform == "Apache2" || .platform == "Nginx" {
+  apache2 = del(.apache2)
+  if is_null(apache2) { apache2 = {} }
+  .http = merge(object!(apache2), {})
+}
+
+
```
Comment (jszwedko) [inline lib/vrl/tests/tests/issues/12948_conditional_type_def_regression.vrl:9]:
```suggestion
```

**Review by jszwedko — COMMENTED** [review]
Nice work! The tests look good. I'll leave it to Jean and Pablo to review the compiler changes.

**Review by JeanMertz — APPROVED** [review]
Sweet 👌 This is a much cleaner approach than the one I was working on, I like it!

**Pull request merged**